### ISSUE TITLE
Video playback within a ListView

### ISSUE THREAD

**Issue by wemped** [flutter/flutter#26767]
A common pattern in apps is having video playing within infinite lists, playing/pausing/stopping as the user scrolls specific views on/off screen. Examples would be the muted autoplay videos on Facebook or Instagram's main feeds.

Best case scenario I think a pattern like this would be a good candidate for one the cookbooks on  https://flutter.io/docs/cookbook
Additionally, I would love to get some insight/suggestions from the community about implementing this kind of feature.

Building similar things on native iOS/Android I would just
1) listen for scroll events
2) check which cells are visible
3) do some math to decide which item should play
4) and push a video player onto that item and play it.

As far as I know - Flutter doesn't have anyway of telling me which Widgets in a ListView are the ones currently visible. The only solution I could come up with was
1. putting a key on every item in the List View
2. using the [https://pub.dartlang.org/packages/rect_getter](RectGetter) package and getting the rect of the ListView
3. get the rect of each item and check if it's within the rect of the ListView

If there's a better way to do this, I'd love to know. I know ListView must be keeping track of what it has rendered somewhere, I just don't know how to access it.

Thanks in advance

**Comment by emailsubjekt** [comment]
I sign this request. This dude is right.

**Comment by wemped** [comment]
@emailsubjekt please add a thumbs up reaction if you agree - the Flutter team prioritizes issues with more thumbs up

**Comment by vijayawoshe** [comment]
any updates on this ? we need this feature

**Comment by wemped** [comment]
Haven't tested this out yet, but came across this widget today, figured it could be useful in when a widget is viewable 

https://github.com/google/flutter.widgets/tree/master/lib/src/visibility_detector

**Comment by yuridiniz** [comment]
For now, this can help us
(https://github.com/rvamsikrishna/inview_notifier_list/blob/master/README.md)[https://github.com/rvamsikrishna/inview_notifier_list/blob/master/README.md]

**Comment by GustavoContreiras** [comment]
I want to display videos from Firebase Storage into a ListView that is on a page that keeps it's state with AutomaticKeepAliveClientMixin.

Each item of the ListView that will hold the VideoPlayer is a custom stateful widget and has SingleTickerProviderStateMixin and AutomaticKeepAliveClientMixin.

Sometimes when I initialize/play/pause a video from network and then try to open ImagePicker
to record a video, at the moment I tap the record icon it crashes with a dialog "Error recording".

In other tries, the state was not kept and the video was download every time I scrolled the ListView.

So, how should I do that? 
- Should I initialize all videos on initState of the page that is holding the ListView?
- Should I initialize the videos only when I tap "play" (not the best approach cause Instagram and Facebook has automatic play when scrolling through it)?
- Should I use a unique global VideoPlayerController to all videos and switch it's value according to the current widget being displayed?
- Should I "uninitialize" the VideoPlayerController after use?

I thought on using CarouselSlider plugin that gives you a custom ListView with a controller that holds the current page/item and then setting a global VideoPlayerController according to the current widget.

Any other thoughts or tips?

**Comment by wemped** [comment]
@GustavoContreiras Your problem might make sense as a separate issue, especially if you're trying to find the root of your crash.

**Comment by Sp4Rx** [comment]
Any update on this issue? @GustavoContreiras

**Comment by GustavoContreiras** [comment]
> 
> 
> Any update on this issue? @GustavoContreiras

I don't know if I'm still getting this "Error recording" issue. I made many revisions to modularize my project and things are working how it should (auto playing videos when they appear on screen and pausing when they dont appear), but I didn't tested enough. It may be a problem of the Samsung device.

I'm using in_view_notifier_list and carousel_slider plugin. Initializing the video on InitState() and playing/pausing on DidUpdateWidget() and OnPageChanged (a method you create for the CarouselSlider widget).

**Comment by lifenautjoe** [comment]
Isn't there a performance issue when using the inview_notifier package? As its a regular `ListView` instead of a `ListView.builder` ?

**Comment by rvamsikrishna** [comment]
@lifenautjoe inview_notifier uses a `ListView.custom`. It does not use a normal `ListView`

**Comment by Legendree** [comment]
Would love to know if someone found a efficient way to do this

**Comment by vamsi-kleargroup** [comment]
@Legendree Check out this package here: https://pub.dev/packages/inview_notifier_list

**Comment by Legendree** [comment]
> @Legendree Check out this package here: https://pub.dev/packages/inview_notifier_list

Yes thank you, I have managed to pull this off with this package.

**Comment by vercorro** [comment]
> A common pattern in apps is having video playing within infinite lists, playing/pausing/stopping as the user scrolls specific views on/off screen. Examples would be the muted autoplay videos on Facebook or Instagram's main feeds.
> 
> Best case scenario I think a pattern like this would be a good candidate for one the cookbooks on https://flutter.io/docs/cookbook
> Additionally, I would love to get some insight/suggestions from the community about implementing this kind of feature.
> 
> Building similar things on native iOS/Android I would just
> 
> 1. listen for scroll events
> 2. check which cells are visible
> 3. do some math to decide which item should play
> 4. and push a video player onto that item and play it.
> 
> As far as I know - Flutter doesn't have anyway of telling me which Widgets in a ListView are the ones currently visible. The only solution I could come up with was
> 
> 1. putting a key on every item in the List View
> 2. using the [https://pub.dartlang.org/packages/rect_getter](RectGetter) package and getting the rect of the ListView
> 3. get the rect of each item and check if it's within the rect of the ListView
> 
> If there's a better way to do this, I'd love to know. I know ListView must be keeping track of what it has rendered somewhere, I just don't know how to access it.
> 
> Thanks in advance

I'm trying to achieve the same thing but my problem is the memory leak of videoPlayer. After several plays/pause (15-20) , videoPlayer crashes. You were trying to achieve the same thing but your issue was how to detect visible widget , what about how you handled the videoPlayer in listView? Please help :)

**Comment by Leffe108** [comment]
For help with how to dispose the video you can look here:
https://stackoverflow.com/questions/58955831/flutter-video-player-dispose

I currently use the solution a bit further down there where I use UniqueKey to dispose the container widget of video_player and replace it with a new one (different UniqueKey) when source url changes. It works solidly in my app where user plays many short videos. 

But I do get a brief empty screen when videos change and need to find solutions to that. So if you get the top solution in the link to work for you, I would recommend to stay with that.

**Comment by taylanyildiz** [comment]
I was searching the same thing and I found this library 
  inview_notifier_list:
    git:
      url: https://github.com/dfdgsdfg/inview_notifier_list.git

**Comment by NaarGes** [comment]
Isn't there any solution that we can manage with flutter `ListView` or `PageView`? Do we have to use a third package to achieve this behavior?

**Comment by taylanyildiz** [comment]
Ofcourse you don't must But there are some problems like ListView get build on every position change you have set true keepalive widget and you have to make the mathemetical settings according to the position change.
I think, it would be wiser to use [InViewNotifierList](https://github.com/dfdgsdfg/inview_notifier_list.git)

**Comment by Samnangsokhorn** [comment]
Does in_view_notifier_list available work with horizontal list?

**Comment by stuartmorgan-g** [comment]
Re-routing to `team-codelab` for evaluation, as the core request here is for codelab-like documentation. `video_player` would be a representative example, but the example could apply to any expensive view (including any platform view).

**Comment by domesticmouse** [comment]
I think this question would be better asked on StackOverflow. My dance card for 2025H1 is already full.

**Comment by Piinks** [comment]
I am going to close this issue since it does not appear to be requesting a specific feature in Flutter, and has been referred to StackOverflow by others on the team. Thank you!
